oeps/oep-0012.rst

```diff
@@ -0,0 +1,520 @@
+interfaces must be added directly to edx-platform, even if the feature
+itself lives in a Python library that has been separately installed.
+
+It would be preferable if any feature implemented as a Python library could
+contribute new sections to the Instructor Dashboard through plugins. The
+dashboard would show a list of the feature that can be managed, and then
+clicking on a feature would present the user with the pluggable interface.
+These interfaces could be made available for instructors in the LMS as
+today, or a new instructor portal could be implemented reusing the same
+blocks.
+
+Specification
+=============
+
+Requirements
+------------
+
+What are the requirements for a pluggable user interface?
+
+1. **It should render its content into a section of a page**
+
+   A pluggable user interface should be able to return the HTML, CSS and custom
+   JavaScript necessary to render its content. In addition, it should be able to
+   declare the JavaScript libraries that it depends upon.
+
+   Python and JavaScript APIs will be provided so that the consumer of a
+   component can render it. These APIs will ensure that the necessary libraries
+   have been loaded, and that the necessary CSS and JavaScript files have been
+   added to the page (without duplication).
+
+2. **It should be able to get the data that it needs to render its content**
+
+   A pluggable user interface needs configuration information, e.g. the course
+   key for the course to be rendered. It also needs to be able to make requests
+   for data that will be represented in the user interface. This data might
+   come from the feature itself, or it might be requested through an API.
+
+3. **It should be able to plug in seamlessly into an existing code base**
+
+   A pluggable user interface needs to provide a unique identifier for itself,
+   and it needs to conform to a contract so that it is clear when and how it can
+   be used. It should not require any changes to the client that consumes it,
+   and it should be possible to be uninstalled cleanly without breaking other
+   parts of the code base.
+
+Proposal
+--------
+
+Django apps provide a standard way to add new functionality to Django, allowing
+new pages, models, and REST APIs to be defined. What is missing is the ability
+to define a component that can be rendered into an existing page, or that can
+be fetched over AJAX and dynamically inserted into a page. This proposal is
+to provide this capability through a new concept called Django component views.
+
+A new `DjangoComponentView` class will be implemented as a subclass of
+`DjangoView` that are then declared in `urls.py` as with other views. They do
+not themselves define user-visible pages but instead provide an implementation
+that can render a component into another page. These views have full access to
+Django so they can use standard mechanisms to access any data that they need to
+be able to render themselves.
```

Comment:
It seems like a lot of the benefit of this could be had by simply allowing edX-specific XBlocks to make use of Django (or even just dropping the cross-platform aspiration of XBlocks). Many already do this, but it's considered bad practice. We could re-consider it good practice.